src/api/message_builder/transaction.rs

```diff
@@ -137,9 +139,55 @@ pub async fn sign_transaction(
         )
         .await?;
     let unlock_blocks = UnlockBlocks::new(unlock_blocks)?;
-    let tx_payload = TransactionPayload::new(prepared_transaction_data.essence, unlock_blocks)?;
+    let tx_payload = TransactionPayload::new(prepared_transaction_data.essence.clone(), unlock_blocks)?;
+
+    let (local_time, milestone_index) = message_builder.client.get_time_and_milestone_checked().await?;
+
+    let conflict = verify_semantic(&prepared_transaction_data, &tx_payload, milestone_index, local_time)?;
+
+    if conflict != ConflictReason::None {
+        return Err(Error::TransactionSemantic(conflict));
+    }
 
-    // validate the signatures in the unlock blocks so we don't send invalid transactions
-    verify_unlock_blocks(&tx_payload, input_addresses)?;
     Ok(Payload::Transaction(Box::new(tx_payload)))
 }
+
+// TODO @thibault-martinez: this is very cumbersome with the current state, will refactor.
+/// Verifies the semantic of a prepared transaction.
+pub fn verify_semantic(
+    prepared_transaction_data: &PreparedTransactionData,
+    transaction: &TransactionPayload,
+    milestone_index: u32,
+    local_time: u64,
+) -> crate::Result<ConflictReason> {
+    let transaction_id = transaction.id();
+    let TransactionEssence::Regular(essence) = transaction.essence();
+    let outputs = prepared_transaction_data
+        .input_signing_data_entries
+        .iter()
+        .map(|i| Ok(Output::try_from(&i.output_response.output)?))
+        .collect::<Result<Vec<Output>>>()?;
+    let inputs = prepared_transaction_data
+        .input_signing_data_entries
+        .iter()
+        .zip(outputs.iter())
+        .map(|(i, o)| {
+            Ok((
+                // PANIC: at this point, OutputIds have been validated many times already.
```

Comment:
If it gets read from the file like the offline example, then it might not be validated before and maybe there are also other use cases, so not sure if we should assume this